### DISCUSSION TITLE
Normalization

### DISCUSSION THREAD

**Discussion by OverLordGoldDragon** [kymatio/kymatio#768] [General]
[This paper](https://arxiv.org/pdf/2007.10926.pdf) normalizes coefficients for DNNs as

<img src="https://user-images.githubusercontent.com/16495490/129969378-a1692ceb-53c0-4ffd-8701-a1f5f04f198d.png" height="53">

with `n=1 to N` samples (independent `x`), and `p` are individual scattering paths.

If I understand it correctly, I see a problem: the _relative scaling_ of coefficients changes. What if a dataset has nearly no energy in high frequencies? The normalization would amplify these which lowers SNR:

![image](https://user-images.githubusercontent.com/16495490/129978440-0298f80d-2402-4988-bf45-6110fba03951.png)

Blue = before, orange = after. Path 1 is clearly dominant to path 0, but post-norm they're of ~equal intensity.

Additional problem for 2D convs, the scaling disparity propagates nonlinearly to a 2D kernel - as if we rescaled differently for 1D convs _along time_, which is a ruin. This'd be alleviated via per-`(n2, n1_fr)` norming instead of per-`n1`.

Is below correct, and any thoughts? @lostanlen 

```python
# time-sum | X.shape == (samples, paths, time)
Xs = X.sum(axis=-1, keepdims=True)
# sample-median
mu = np.median(Xs, axis=0, keepdims=True)
# rescale
Xmu = X / mu
# configurable as long as Xmu > -1
eps = np.abs(1.1 * Xmu.min())
Xmu = Xmu / eps
# log
Xmu = np.log(1 + Xmu)
# left eq global time averages in norming but I didn't
```

<details>
  <summary><b>full code</b></summary>

```python
import numpy as np
import scipy.signal as sig
import matplotlib.pyplot as plt
from numpy.fft import fft, ifft, ifftshift
from kymatio.visuals import plot

def normalize(X):
    # time-sum
    Xs = X.sum(axis=-1, keepdims=True)
    # sample-median
    mu = np.median(Xs, axis=0, keepdims=True)
    # rescale
    Xmu = X / mu
    # ensure >-1
    eps = np.abs(1.1 * Xmu.min())
    Xmu = Xmu / eps
    # log
    Xmu = np.log(1 + Xmu)
    return Xmu

def hist(x):
    plt.hist(x.ravel(), bins=500)
    plt.show()

# (N, p, t)
shape = (128, 32, 2048)

#%% generate non-WGN and smooth it for convenience
np.random.seed(0)
X = np.zeros(shape)
shape_h = (*shape[:-1], shape[-1]//2)

X[..., :shape[-1]//2:] = np.random.randn(*shape_h)
X[..., shape[-1]//2:] = np.random.uniform(-3, 3.5, shape_h)
w = ifftshift(sig.gaussian(shape[-1], 8))
X = ifft(fft(X) * fft(w)).real

#%% simulate low-energy path
X[:, 0] /= 20

# normalize
Xmu = normalize(X)

#%%
hist(X[0][0])
hist(Xmu[0][0])
#%%
x00, x01, xmu00, xmu01 = X[0, 0], X[0, 1], Xmu[0, 0], Xmu[0, 1]
plot(x00, title="path 0: before & after norm")
plot(xmu00, show=1)
plot(x01, title="path 1: before & after norm")
plot(xmu01, show=1)
```
</details>

**Comment by lostanlen** [comment]
> the relative scaling of coefficients changes

yes


> What if a dataset has nearly no energy in high frequencies? The normalization would amplify these which lowers SNR

yes


> Path 1 is clearly dominant to path 0, but post-norm they're of ~equal intensity.

yes


> Additional problem for 2D convs, the scaling disparity propagates nonlinearly to a 2D kernel 

yes but the paper you cite isn't about 2D convnets. it's about large-margin nearest neighbors

**Reply by OverLordGoldDragon** [reply]
Is [DSS](https://arxiv.org/pdf/1304.6763.pdf)'s renorm of `S2` by `S1` advised for JTFS? Unsure if theorems persist past frequential scattering.

**Reply by lostanlen** [reply]
Short answer is no. We didn't renormalize in the JTFS paper, nor (as far as i remember) in any of the following ones with Joakim or Changhong or Florian.

The reason why we didn't do it is that if the input is well-localized in frequency (unique exponential chirp for example), then there are some coefficients `S2` which are large even though `S1` is tiny. This arises near the chirp (not on the chirp) if you don't apply `average_fr` to `S1`. So this may cause instabilities because these near-chirp paths cause `S2/S1` to take up most of the representation's norm, and thus to account for most of the Euclidean distance between signals.

Of course, this can be mitigated by averaging `S1` over the frequency domain, thus "spreading" the chirp over the frequency domain and avoiding tiny denominators in the renormalization.

But the solution we opted for is simpler: logarithmic transformation. `S1(lambda) ---> log1p(S1(lambda)/mu(lambda))` and likewise for `S2`, where `mu(lambda)` is a path-dependent (time-independent) renormalization constant which can be learned in a supervised or unsupervised way on the training set set.
(i think i did something like `mu(lambda) = median(S(lambda))/1000` or something like that. this is just a heuristic, there are better ones)

Note that log transformation achieves something close to renormalization, because `log(A/B) = log(A) - log(B)`. So any linear transformation after the log (SVM, LMNN, convnet) will be capable to perform an `S2/S1`-like renormalization if that's helpful to the task.

**Comment by OverLordGoldDragon** [comment]
To counter relative scaling disparity, I introduce a post-processing step and change an existing one:

 1. Scale paths by _original_ norms (L1 or L2, used L1 below)
 2. Standardize per _sample_ statistics, rather than per path. This requires computing `mean` and `std` for each new input, but is the only way to avoid relative scaling along features or spatial dimensions.
 3. In case of 2D convs, a "path" will be considered a joint slice, i.e. `(n2, n1_fr)`, rather than `n1`. 3D convs will adjust likewise; idea's to standardize dimensions which are convolved over with same `std` and `mean`.

These largely retain Gaussianization in examples I've tried:

![image](https://user-images.githubusercontent.com/16495490/131236867-0b5f4df9-3dda-46f4-ab68-3fb9afab0576.png)

compared with original:

![image](https://user-images.githubusercontent.com/16495490/131237398-fded7038-c44e-4a1d-8e2d-38a65fb7cf0f.png)

So "new" differs mainly with the distribution spike due to retaining input's scaling.

There's also the option to keep relative _log_-scalings, that is, maintain logs of ratios of norms between paths - but this becomes dependent on log base.

<details>
  <summary><b>code</b></summary>

```python
import numpy as np
import scipy.signal as sig
from numpy.fft import fft, ifft, ifftshift
from kymatio.visuals import plot, hist

#%%
def normalize(X, std_axis=(1, 2), mean_axis=(1, 2), rscaling='l1', C=None):
    # main transform #########################################################
    # time-sum
    Xs = X.sum(axis=-1, keepdims=True)
    # sample-median
    mu = np.median(Xs, axis=0, keepdims=True)
    # rescale
    Xmu = X / mu
    # log
    if C is None:
        C = 130 * np.abs(X).mean()
    Xmu = np.log(1 + Xmu * C)
    Xnorm = Xmu
        
    # standardization + relative scaling #####################################
    if mean_axis is not None:
        Xnorm -= Xnorm.mean(axis=mean_axis, keepdims=True)
    
    if rscaling is not None:
        kw = dict(axis=-1, keepdims=True)
        ord = (2 if rscaling == 'l2' else 1)
        Xstd = X - X.mean(**kw)
        norms_orig = np.linalg.norm(Xstd,  ord=ord, **kw)
        norms_now  = np.linalg.norm(Xnorm, ord=ord, **kw)
        norms = norms_orig / norms_now
        Xnorm *= (norms / norms.mean())
    
    if std_axis is not None:
        Xnorm /= Xnorm.std(axis=std_axis, keepdims=True)
    return Xnorm

#%% generate non-WGN and smooth it for convenience
# (N, p, t)
shape = (512, 50, 129)

np.random.seed(0)
X = np.zeros(shape)
N3 = shape[-1]//3
shape_h = (*shape[:-1], N3)

X[..., :N3] = np.random.randn(*shape_h)
X[..., N3:2*N3] = np.random.uniform(-4, 1, shape_h)
X[..., 2*N3:] = np.random.wald(3, 5, shape_h)
w = ifftshift(sig.gaussian(shape[-1], 2))
w[np.abs(w) < 1e-8] = 0
X = ifft(fft(X) * fft(w)).real

X[:, 0] /= 20
X = np.abs(X)
#%% simulate low-energy path
Xmu = normalize(X)
x00, x01, xmu00, xmu01 = X[0, 0], X[0, 1], Xmu[0, 0], Xmu[0, 1]

#%%
hist(X,   show=1, title="hist(X), before")
hist(Xmu, show=1, title="hist(X), after", xlims=(-3.5, 3.5))
#%%
plot(x00, title="path 0: before & after norm")
plot(xmu00, show=1)
plot(x01, title="path 1: before & after norm")
plot(xmu01, show=1)
#%%
plot(x00,   title="path 0: before", show=1)
plot(xmu00, title="path 0: after",  show=1)
#%%
plot(x01,   title="path 1: before", show=1)
plot(xmu01, title="path 1: after",  show=1)
#%%
Xmu = normalize(X, rscaling=None, mean_axis=(0, 2), std_axis=(0, 2))
hist(Xmu, show=1, title="hist(X), after", xlims=(-3.5, 3.5))
```
</details>

**Reply by lostanlen** [reply]
i don't understand what problem you're solving or trying to solve. You seem to assume that changing the relative scaling of coefficients necessarily harms the ability of convnets to learn. The success of PCEN (https://arxiv.org/abs/1607.05666) is a counterexample to that. I'm not saying path normalization always helps, but at least it shows that the objection you're raising ("what if a dataset has nearly no energy in high frequencies?") isn't as clear-cut.

Have you trained a convnet on a task and observed something particular or are you speaking in general?

If, as you describe, a dataset has nearly no energy in high frequencies, the solution is to resample the input to a lower sample rate so that the scattering paths corresponding to these frequencies disappear from the scattering network. This is numerically stabler and faster than any post-hoc renormalization technique.

Now, if you're right and normalization harms convnet learning (which is easy to imagine), then by all means, disable it. What i don't understand is that you seem to assume that one normalization of some kind must happen, just not one we published, so you invent another. But first of all, why do you think that one normalization of some kind must happen? Isn't batch-norm enough?

The replacement you propose ("Standardize per sample statistics, rather than per path") is not really a replacement because it doesn't satisfy at all the same properties.
If you standardize per sample statistics, then `x` and `k*x` (with `k` a constant) are mapped to the same coefficients. This could perhaps be useful if there is some uncontrolled variation in gain during acquisition. But it could also easily be addressed by resorting to cosine similarity instead of Euclidean similarity (i can explain in detail if you wish). So i don't necessarily see a use.
On the other hand, renormalizing per path will not cause `x` and `kx` to be mapped to the same coefficients. It is more akin to PCEN. Its role is to renormalize the effect of path-dependent stationary noise (which is very frequent in acoustics and biomedical signal processing). 

There is not and will not be a cookie-cutter solution. Every pipeline has its own built-in assumptions (useful expressed in terms of invariance/equivariance/discriminability) so it will always be possible to devise a dataset which defeats those assumptions. In my opinion, the crux of the matter is to understand how to reduce intra-class variability in real datasets. Sample normalization and path normalization are two very different techniques to achieve this. They are not comparable except superficially in the sense that something gets divided by something else in the hope that a multiplicative factor will cancel out in the quotient.

**Reply by OverLordGoldDragon** [reply]
My problem with existing norm is two-fold:

 1. Amplifying noise in some samples because others have signal there
 2. Breaking spatial coherence necessary for higher-dimensional convs (like scale-shifting every individual timestep for 1D convs)

Seems we agree on 2.

> resample the input to a lower sample rate

This tosses information

> normalization of some kind must happen

I don't assume so - I simply know a principled normalization can help significantly and am trying to improve an existing method.

> BN

followed by log if we want log's benefits... sure

> just not one we published

Both implemented; the published variant is accessed with appropriate (and documented) axis kwargs, and they can be made default.

> Have you trained ... or are you speaking in general?

Latter

> PCEN

I have no problem with; it's a _pointwise_ transform, just like `log`, meaning it's determined independently for each sample and doesn't suffer the "amplify channel noise because signal sometimes there". Granted the _learned_ variant does determine parameters per-channel, but the "learned" part should figure out not to amplify noise.

>  2. Breaking spatial coherence necessary for higher-dimensional convs (like scale-shifting every individual timestep for 1D convs)

Compromise is possible: define "channels" as non-conv dims. This makes "path" e.g. `(n1_fr, n1)` rather than just `n1`, so all spatial dims are rescaled the same.

> then `x` and `k*x` are mapped to the same coefficients. 

Isn't this the case with most standardizing? Global scalar scaling doesn't carry meaningful information yet impedes training, so we kill it.

> There is not and will not be a cookie-cutter solution

Right - it's points worth noting.
